toolchain: gcc: adjust configuration

rockdrilla opened the pull request:
- provide sane way for specifying variables for both "configure" and "make"
- unify configuration with different libc implementations

rockdrilla commented:
@robimarko, @hauke, @rsalvaterra, friendly ping, PTAL.

Ansuel commented:
@rockdrilla it's not clear why this is needed?

robimarko commented:
> @rockdrilla it's not clear why this is needed?

I second this, especially since the description does not really say what you are doing and why, and I find it way too vague

rockdrilla commented:
I've observed (GCC, final stage) configuration drift between musl and glibc while working on #18647:
```sh
# musl
lt_cv_prog_compiler_static_works=yes
lt_cv_prog_compiler_static_works_CXX=yes
lt_cv_sys_max_cmd_len=1572864

# glibc
lt_cv_prog_compiler_static_works=no
lt_cv_prog_compiler_static_works_CXX=no
lt_cv_sys_max_cmd_len=512

```


These changes should prevent this issue in future:
```sh
export lt_cv_prog_compiler_static_works=yes
export lt_cv_prog_compiler_static_works_CXX=yes
export lt_cv_sys_max_cmd_len=1572864
```

At last but not least:
- provide custom `autotools`/`libtool` variables via properly named variable (`GCC_CONFIGURE_VARS` ),
- move variables from `GCC_MAKE` to `GCC_CONFIGURE_VARS` (at this moment only `gcc_cv_libc_provides_ssp=yes` for musl),
- propagate it's usage for both `./configure` and `make`.

Ansuel commented:
@rockdrilla ok would be ideal to add those finding in the commit. Maybe a better description would be "make gcc config consistent with glib/musl"

rockdrilla commented:
"_make gcc great again_" (sorry, just kidding)

You're right - I'll add this information to commit message.

Ansuel commented:
![ChatGPT Image 2 mag 2025, 00_03_09](https://github.com/user-attachments/assets/ebcd0720-27d4-4607-81ef-484b193f363b)

rockdrilla committed:
toolchain: gcc: make config consistent with glibc/musl

I've observed configuration drift for GCC between musl and glibc
(especially it's final stage):

  # musl
  lt_cv_prog_compiler_static_works=yes
  lt_cv_prog_compiler_static_works_CXX=yes
  lt_cv_sys_max_cmd_len=1572864

  # glibc
  lt_cv_prog_compiler_static_works=no
  lt_cv_prog_compiler_static_works_CXX=no
  lt_cv_sys_max_cmd_len=512

These changes should prevent this issue in future:

  export lt_cv_prog_compiler_static_works=yes
  export lt_cv_prog_compiler_static_works_CXX=yes
  export lt_cv_sys_max_cmd_len=1572864

Also:

- provide custom autotools/libtool variables via properly named
  variable ("GCC_CONFIGURE_VARS"),
- move variables from "GCC_MAKE" to "GCC_CONFIGURE_VARS"
  (at this moment only "gcc_cv_libc_provides_ssp=yes" for musl),
- propagate it's usage for both "./configure" and "make".

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18646
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

robimarko commented:
This completely broke all buildbots and local building as well for the GCC initial configuration so I reverted it.

Please post a fixed version and test thoroughly locally before.

rockdrilla commented:
It looks like `toolchain/gdb` failed somehow. IDK why `toolchain/gcc` affects it.

robimarko commented:
It fails on GCC initial config, not GDB, example:
```
bash: -c: line 1: syntax error near unexpected token `;'
bash: -c: line 1: `(cd /__w/openwrt/openwrt/openwrt/build_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.2.0_musl_eabi/gcc-14.2.0-initial && rm -f config.cache; ; SHELL="bash" /__w/openwrt/openwrt/openwrt/build_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.2.0_musl_eabi/gcc-14.2.0/configure --with-bugurl=http://bugs.openwrt.org/ --with-pkgversion="OpenWrt GCC 14.2.0 r0-57841c8" --prefix=/__w/openwrt/openwrt/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.2.0_musl_eabi --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --target=arm-openwrt-linux-muslgnueabi --with-gnu-ld --enable-target-optspace --disable-libgomp --disable-libmudflap --disable-multilib --disable-libmpx --disable-nls --disable-libssp --without-isl --without-cloog --with-host-libstdcxx=-lstdc++      --with-system-zlib=/__w/openwrt/openwrt/openwrt/staging_dir/host --with-zstd=/__w/openwrt/openwrt/openwrt/staging_dir/host --with-gmp=/__w/openwrt/openwrt/openwrt/staging_dir/host --with-mpfr=/__w/openwrt/openwrt/openwrt/staging_dir/host --with-mpc=/__w/openwrt/openwrt/openwrt/staging_dir/host --disable-decimal-float --with-diagnostics-color=auto-if-env --enable-__cxa_atexit --enable-libstdcxx-dual-abi --with-default-libstdcxx-abi=new --with-cpu="cortex-a9" --with-fpu="vfpv3-d16" --with-float=hard CFLAGS="-O2 -I/__w/openwrt/openwrt/openwrt/staging_dir/host/include  -pipe" CXXFLAGS="-O2 -I/__w/openwrt/openwrt/openwrt/staging_dir/host/include  -pipe" CFLAGS_FOR_TARGET="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" CXXFLAGS_FOR_TARGET="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" GOCFLAGS_FOR_TARGET="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" --with-newlib --with-sysroot=/__w/openwrt/openwrt/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.2.0_musl_eabi --enable-languages=c --disable-shared --disable-threads  );'
make[2]: *** [Makefile:31: /__w/openwrt/openwrt/openwrt/build_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.2.0_musl_eabi/gcc-14.2.0-initial/.configured] Error 2
time: toolchain/gcc/initial/compile#5.96#3.62#7.52
```

rockdrilla commented:
I've found that error and fixed that:
```diff
--- a/toolchain/gcc/common.mk
+++ b/toolchain/gcc/common.mk
@@ -229,7 +229,7 @@ endif
 
 define Host/Configure
        (cd $(GCC_BUILD_DIR) && rm -f config.cache; \
-               $(if $(strip $(GCC_CONFIGURE_VARS)),export $(GCC_CONFIGURE_VARS)); \
+               $(if $(strip $(GCC_CONFIGURE_VARS)),export $(GCC_CONFIGURE_VARS);) \
                $(GCC_CONFIGURE) \
        );
 endef
--- a/toolchain/gcc/final/Makefile
+++ b/toolchain/gcc/final/Makefile
@@ -39,7 +39,7 @@ define Host/Configure
        $(CleanupToolchain)
        mkdir -p $(GCC_BUILD_DIR)
        (cd $(GCC_BUILD_DIR) && rm -f config.cache; \
-               $(if $(strip $(GCC_CONFIGURE_VARS)),export $(GCC_CONFIGURE_VARS)); \
+               $(if $(strip $(GCC_CONFIGURE_VARS)),export $(GCC_CONFIGURE_VARS);) \
                $(GCC_CONFIGURE) \
        );
 endef
```

Would you accept that change?

robimarko commented:
Please make a PR
